pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,458 @@
+package shortnames
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/manifoldco/promptui"
+	"github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// IsShortName returns true if the specified input is a "short name".  A "short
+// name" refers to a container image without a fully-qualified reference, and
+// is hence missing a registry (or domain).  Names including a digest are not
+// short names.
+//
+// Examples:
+//  * short names: "image:tag", "library/fedora"
+//  * not short names: "quay.io/image", "localhost/image:tag",
+//                     "server.org:5000/lib/image", "image@sha256:..."
+func IsShortName(input string) bool {
+	isShort, _, _ := parseUnnormalizedShortName(input)
+	return isShort
+}
+
+// parseUnnormalizedShortName parses the input and returns if it's short name,
+// the unnormalized reference.Named, and a parsing error.
+func parseUnnormalizedShortName(input string) (bool, reference.Named, error) {
+	ref, err := reference.Parse(input)
+	if err != nil {
+		return false, nil, errors.Wrapf(err, "cannot parse input: %q", input)
+	}
+
+	named, ok := ref.(reference.Named)
+	if !ok {
+		return true, nil, errors.Errorf("%q is not a named reference", input)
+	}
+
+	registry := reference.Domain(named)
+	if strings.ContainsAny(registry, ".:") || registry == "localhost" {
+		// A final parse to make sure that docker.io references are correctly
+		// normalized (e.g., docker.io/alpine to docker.io/library/alpine.
+		named, err = reference.ParseNormalizedNamed(input)
+		if err != nil {
+			return false, nil, errors.Wrapf(err, "cannot normalize input: %q", input)
+		}
+		return false, named, nil
+	}
+
+	return true, named, nil
+}
+
+// splitUserInput parses the user-specified reference.  Namely, it strips off
+// the tag or digest and stores it in the return values so that both can be
+// re-added to a possible resolved alias' or USRs at a later point.
+func splitUserInput(named reference.Named) (isTagged bool, isDigested bool, normalized reference.Named, tag string, digest digest.Digest) {
+	normalized = named
+
+	tagged, isT := named.(reference.NamedTagged)
+	if isT {
+		isTagged = true
+		tag = tagged.Tag()
+	}
+
+	digested, isD := named.(reference.Digested)
+	if isD {
+		isDigested = true
+		digest = digested.Digest()
+	}
+
+	// Strip off tag/digest if present.
+	normalized = reference.TrimNamed(named)
+
+	return
+}
+
+// Add records the specified name-value pair as a new short-name alias to the
+// user-specific aliases.conf.  It may override an existing alias for `name`.
+func Add(ctx *types.SystemContext, name reference.Named, value reference.Named) error {
```

Comment:
Having a `value reference.Named` is nice; using `reference.Named` for the alias loses the earlier-mentioned separation of short names vs. normalized names via the type system.

We didn’t get to a full type-system separation, still…

Could the alias name be a `string` again, please?

Same for `Remove`.

Reply:
done